dnn build fails using dispatch mode

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => master ( ce6b06efb9fecf3efadf6c1c64c5b687c937341e )
- Operating System / Platform => Win 7 Pro 64bit
- Compiler => Visual Studio 2013 Update 4

##### Detailed description

 - I tried to compile recent git version OpenCV on VS2013
 - I modified following parts
   - ```CPU_BASELINE```:```AVX```
   - ```CPU_DISPATCH```:```AVX2```
   - other minor modifications
 - When I built ```dnn``` module, I get dozen of following link error
```
convolution_layer.obj : error LNK2005: "void __cdecl cv::dnn::cpu_baseline::fastConv(float const *,unsigned __int64,float const *,float const *,float *,int const *,int,int,int,float const *,bool)" (?fastConv@cpu_baseline@dnn@cv@@YAXPEBM_K00PEAMPEBHHHH0_N@Z) already defined in concat_layer.obj
convolution_layer.obj : error LNK2005: "void __cdecl cv::dnn::cpu_baseline::fastGEMM1T(float const *,float const *,unsigned __int64,float const *,float *,int,int)" (?fastGEMM1T@cpu_baseline@dnn@cv@@YAXPEBM0_K0PEAMHH@Z) already defined in concat_layer.obj
convolution_layer.obj : error LNK2005: "void __cdecl cv::dnn::cpu_baseline::fastGEMM(float const *,unsigned __int64,float const *,unsigned __int64,float *,unsigned __int64,int,int,int)" (?fastGEMM@cpu_baseline@dnn@cv@@YAXPEBM_K01PEAM1HHH@Z) already defined in concat_layer.obj
crop_layer.obj : error LNK2005: "void __cdecl cv::dnn::cpu_baseline::fastConv(float const *,unsigned __int64,float const *,float const *,float *,int const *,int,int,int,float const *,bool)" (?fastConv@cpu_baseline@dnn@cv@@YAXPEBM_K00PEAMPEBHHHH0_N@Z) already defined in concat_layer.obj
crop_layer.obj : error LNK2005: "void __cdecl cv::dnn::cpu_baseline::fastGEMM1T(float const *,float const *,unsigned __int64,float const *,float *,int,int)" (?fastGEMM1T@cpu_baseline@dnn@cv@@YAXPEBM0_K0PEAMHH@Z) already defined in concat_layer.obj
crop_layer.obj : error LNK2005: "void __cdecl cv::dnn::cpu_baseline::fastGEMM(float const *,unsigned __int64,float const *,unsigned __int64,float *,unsigned __int64,int,int,int)" (?fastGEMM@cpu_baseline@dnn@cv@@YAXPEBM_K01PEAM1HHH@Z) already defined in concat_layer.obj
(snip)
build\bin\Release\opencv_dnn330.dll : fatal error LNK1169: one or more multiply defined symbols found
```

 - Probably we need to add some include guard in correct place

##### Steps to reproduce

 - Based on cmake output, I modified following options
```
-DWITH_IPP:BOOL="0"
-DCPU_DISPATCH:STRING="AVX2"
-DWITH_CUDA:BOOL="0"
-DCPU_BASELINE:STRING="AVX"
-DANT_EXECUTABLE:FILEPATH="ANT_EXECUTABLE-NOTFOUND"
-DWITH_OPENCL:BOOL="0" 
```
 - Probably the important option is ```CPU_DISPATCH``` and ```CPU_BASELINE```

--- a/modules/dnn/src/layers/layers_common.hpp
+++ b/modules/dnn/src/layers/layers_common.hpp
@@ -45,9 +45,11 @@
 #include <opencv2/dnn.hpp>
 #include <opencv2/dnn/shape_utils.hpp>
 
+#define CV_CPU_OPTIMIZATION_DECLARATIONS_ONLY
 // dispatched AVX/AVX2 optimizations
 #include "layers/layers_common.simd.hpp"
 #include "layers/layers_common.simd_declarations.hpp"
+#undef CV_CPU_OPTIMIZATION_DECLARATIONS_ONLY
 
 namespace cv
 {

